src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -823,26 +825,53 @@ private async Task CreateServicesAsync(CancellationToken cancellationToken = def
                 return;
             }
 
-            // We do not specify the initial list version, so the watcher will give us all updates to Service objects.
-            IAsyncEnumerable<(WatchEventType, Service)> serviceChangeEnumerator = kubernetesService.WatchAsync<Service>(cancellationToken: cancellationToken);
-            await foreach (var (evt, updated) in serviceChangeEnumerator)
-            {
-                if (evt == WatchEventType.Bookmark) { continue; } // Bookmarks do not contain any data.
+            const int watchAttempts = 2;
 
-                var srvResource = needAddressAllocated.Where(sr => sr.Service.Metadata.Name == updated.Metadata.Name).FirstOrDefault();
-                if (srvResource == null) { continue; } // This service most likely already has full address information, so it is not on needAddressAllocated list.
+            for (int attempt = 0; attempt < watchAttempts; attempt++)
+            {
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                cts.CancelAfter(configuration.GetValue<TimeSpan>("DOTNET_ASPIRE_SERVICE_STARTUP_WATCH_TIMEOUT", TimeSpan.FromSeconds(10)));
 
-                if (updated.HasCompleteAddress || updated.Spec.AddressAllocationMode == AddressAllocationModes.Proxyless)
+                try
                 {
-                    srvResource.Service.ApplyAddressInfoFrom(updated);
-                    needAddressAllocated.Remove(srvResource);
+                    // We do not specify the initial list version, so the watcher will give us all updates to Service objects.
+                    IAsyncEnumerable<(WatchEventType, Service)> serviceChangeEnumerator = kubernetesService.WatchAsync<Service>(cancellationToken: cts.Token);
+                    await foreach (var (evt, updated) in serviceChangeEnumerator)
+                    {
+                        if (evt == WatchEventType.Bookmark) { continue; } // Bookmarks do not contain any data.
+
+                        var srvResource = needAddressAllocated.FirstOrDefault(sr => sr.Service.Metadata.Name == updated.Metadata.Name);
+                        if (srvResource == null) { continue; } // This service most likely already has full address information, so it is not on needAddressAllocated list.
+
+                        if (updated.HasCompleteAddress)
+                        {
+                            srvResource.Service.ApplyAddressInfoFrom(updated);
+                            needAddressAllocated.Remove(srvResource);
+                        }
+
+                        if (needAddressAllocated.Count == 0)
+                        {
+                            return; // We are done
+                        }
+                    }
                 }
+                catch { }
```

Comment:
Can we be more specific with the catch here? Do we know what specific exceptions are likely to occur that we can retry and everything else is unexpected and therefore fatal? We should be looking to at least log these exceptions via `ILogger.LogDebug`.

Reply:
done